contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,853 @@
+
+1. If the StatefulSet's `.Spec.TemplateGeneration` is nil, the StatefulSet 
+has never been updated, and its history has never been initialized. This is 
+the state the object will be in when a cluster is first upgraded from a version
+that does not support StatefulSet update to a version that does. In this case,
+the controller will not enforce PodTemplate revisions. When creating Pods, 
+it will always use the StatefulSet's `.Spec.Template`. Otherwise, the controller 
+will continue as below.
+1. The controller will select all PodTemplates with a 
+`StatefulSetPodTemplateLabel` matching the `.Name` field of the StatefulSet.
+1. The controller will filter out all PodTemplates that do not contain a 
+ControllerRef matching the the StatefulSet. If the controller selects 
+PodTemplates that it does not own, it will report an error, but it will continue
+reconstructing the StatefulSet's history.
+1. The controller will filter out all PodTemplates that do not have a 
+`StatefulSetTemplateGenerationLabel` mapped to a valid revision. This can only 
+occur if the user purposefully deletes the label. In this case, the 
+controller will report an error, but it will continue reconstructing the 
+StatefulSet's revision history.
+1. For all the remaining PodTemplates, the controller will sort them in 
+ascending order by the value mapped to their `StatefulSetTemplateGenerationLabel`. 
+This will reconstruct a list of PodTemplates from oldest to newest. Note that, 
+as the revision is monotonically increasing for an individual StatefulSet, and
+as we use ControllerRef to mitigate selector overlap, the StatefulSet's history
+ is a strictly ordered set.
+
+#### History Maintenance 
+In order to prevent the revision history of the StatefulSet from exceeding 
+memory or storage limits, the StatefulSet controller will periodically prune 
+the oldest PodTemplates from the StatefulSet's revision history. 
+
+1. The StatefulSet controller will 
+[reconstruct the revision history](#history-reconstruction) 
+of the StatefulSet.
+1. If the number of PodTemplates in the StatefulSet's revision history is 
+greater than the StatefulSet's `.Spec.RevisionHistoryLimit`, the 
+StatefulSet controller will delete PodTemplates, starting with the head of 
+the revision history, until the limit of the revision history is equal to 
+the StatefulSet's `.Spec.RevisionHistoryLimit`.
+1. As a StatefulSet's `.Spec.RevisionHistoryLimit` is always at least two, and 
+as the PodTemplates corresponding to `.Status.TemplateGeneration` 
+or `.Status.TargetTemplateGeneration` are always the most recent PodTemplates 
+in the revision history, the StatefulSet controller will not delete any 
+`PodTemplates` that represent the current or target revisions.
+
+### Template Updates
+The StatefulSet controller will create PodTemplates upon mutation of the 
+`.Spec.Template` of a StatefulSet.
+
+1. When the StafefulSet controller observes a mutation to a StatefulSet's 
```

Comment:
It will probably be cheaper to simply compare .Spec.TemplateGeneration with .Status.TargetTemplateGeneration rather than deepEqual between statefulset updates.